cfgrammar/src/lib/span.rs

```diff
@@ -46,14 +48,20 @@ impl Span {
 
 pub struct NewlineToLineColCache {
     newlines: Vec<usize>,
+    src_pos: usize,
     trailing_bytes: usize,
+    trailing_chars: usize,
+    trailing: HashMap<usize, (usize, usize)>,
 }
 
 impl Default for NewlineToLineColCache {
     fn default() -> NewlineToLineColCache {
         NewlineToLineColCache {
             newlines: vec![0],
             trailing_bytes: 0,
+            trailing_chars: 0,
+            trailing: HashMap::new(),
```

Comment:
What are `trailing_chars` and (in particular) `trailing`?

Reply:
`trailing_chars` and `trailing_bytes` are accumulators, where `trailing` is a "snapshot" of `(trailing_bytes, trailing_chars)` at a point where feed was called, the key of trailing is the `src_pos` of the *next* call to feed beyond the current one.

`trailing_chars`, is just the number of characters trailing the previous newline, we can't calculate it from bytes because of unicode characters, the copy in trailing is used to calculate the column in cases where the newline is in a previous string than the one `byte_to_line_and_col` is passed as `src`. The copy of `trailing_bytes` in `trailing` is used similarly to slice into the `src` string passed into `bytes_to_line_and_col`.  We need a snapshot of each `trailing_bytes` and `trailing_chars` for each potential call to `bytes_to_line_and_col`

for example:
`cache.feed("\nabc"); cache.feed("mnop"); cache.feed("xyz");`

`trailing` should (I believe be) `{ 4 = (3,3), 8 = (7,7)}`, We could add the first `0 = (0, 0)`, but it is unnecessary the previous has never has trailing bytes or chars.
With 4 being the `src_pos` for `"mnop"` and 8 for `"xyz"`, we can then calculate that `'n'` of `"mnop"` is at the 5th column, even though `"abc"` was never passed to `bytes_to_line_and_col`.

In these ASCII cases they'll be equal, but should differ for unicode (another test).

Reply:
Is the idea here that we want to cope with strings that end in non-valid-unicode characters?

Reply:
No, just valid multibyte where each one which only add `1` to the column offset, `trailing_bytes` part of this is an artifact of the way we know we are slicing at a unicode `char_boundary` by doing so after the newline we know it is a valid `char_boundary`.  

Your questions trigger me to think that we can perhaps get rid of this usage of `trailing_bytes` though.
`if line_off < src_pos` my feeling is I'm probably doing a bunch of extra work to calculate `src[0..]`, aka `src` given a `line_off` by subtracting out stuff...

Reply:
Indeed, now that I have that one major edge case under wraps, I will add unicode tests, test the `None` cases, and some more general testing, and stuff at `EOF`.  But I don't know of any more major edge cases lurking, but no proof they don't exist...